SYCL/ESIMD/api/functional/ctors/ctor_move.cpp

```diff
@@ -0,0 +1,232 @@
+          passed = false;
+
+          log::fail(ctors::TestDescription<DataT, NumElems, TestCaseT>(
+              i, retrieved, expected, data_type));
+        }
+      }
+    }
+
+    return passed;
+  }
+
+private:
+  // The core check logic.
+  // Uses USM pointers for input data and to store the data from the new simd
+  // instance, so that we could check it later
+  // Returns the flag that should be true only if the move constructor was
+  // actually called, to differentiate with the copy constructor calls
+  static bool run_check(const DataT *const in, DataT *const out) {
+    bool was_moved = false;
+
+    // Prepare the source simd to move
+    simd<DataT, NumElems> source;
+    source.copy_from(in);
+
+    // Action to run over the simd move constructor result
+    const auto action = [&](const simd<DataT, NumElems> &instance) {
+      was_moved = instance.get_test_proxy().was_move_destination();
+      instance.copy_to(out);
+    };
+    // Call the move constructor in the specific context and run action
+    // directly over the simd move constructor result
+    TestCaseT::template run(std::move(source), action);
+
+    return was_moved;
+  }
+};
+
+// Uses the initializer C++ context to call simd move constructor
+struct initializer {
+  static std::string get_description() { return "initializer"; }
+
+  template <typename SimdT, typename ActionT>
+  static void run(SimdT &&source, const ActionT &action) {
+    static_assert(
+        type_traits::is_nonconst_rvalue_reference_v<decltype(source)>);
+
+    const auto instance = SimdT(std::move(source));
+    action(instance);
+  }
+};
+
+// Uses the variable declaration C++ context to call simd move constructor
+struct var_decl {
+  static std::string get_description() { return "variable declaration"; }
+
+  template <typename SimdT, typename ActionT>
+  static void run(SimdT &&source, const ActionT &action) {
+    static_assert(
+        type_traits::is_nonconst_rvalue_reference_v<decltype(source)>);
+
+    const auto instance(std::move(source));
+    action(instance);
+  }
+};
+
+// Uses the rvalue in expression C++ context to call simd move constructor
+struct rval_in_expr {
+  static std::string get_description() { return "rvalue in expression"; }
+
+  template <typename SimdT, typename ActionT>
+  static void run(SimdT &&source, const ActionT &action) {
+    static_assert(
+        type_traits::is_nonconst_rvalue_reference_v<decltype(source)>);
+
+    SimdT instance;
+    instance = SimdT(std::move(source));
+    action(instance);
+  }
+};
+
+// Uses the function argument C++ context to call simd move constructor
+class const_ref {
+public:
+  static std::string get_description() { return "const reference"; }
+
+  template <typename SimdT, typename ActionT>
+  static void run(SimdT &&source, const ActionT &action) {
+    static_assert(
+        type_traits::is_nonconst_rvalue_reference_v<decltype(source)>);
+
+    action(SimdT(std::move(source)));
+  }
+};
```

Comment:
thinks about moving struct, that defined contexts, into the top of this file

Reply:
Done by 51a796566cff3d78d86b8dc1929dd336308e73be